File: server/evaluation/wrappers/baseline-calculator.ts
/**
 * A2: Safety Baseline Calculator - TypeScript Wrapper
 *
 * Phase 5: Refactored to use real AppsFlyer cohort data.
 * The Python-based calculation is kept for backward compatibility but deprecated.
 */

import { spawn } from 'child_process'
import path from 'path'
import { getBaselineMetrics } from '../../db/queries-appsflyer'
import {
  getBaselineSettings,
  upsertBaselineSettings,
  upsertSafetyBaseline,
  getSafetyBaseline,
} from '../../db/queries-evaluation'
import type { BaselineSettings, NewBaselineSettings } from '../../db/schema'

export interface BaselineResult {
  baseline_roas7: number | null
  baseline_ret7: number | null
  reference_period: string
  total_spend?: number
  total_revenue?: number
  total_installs?: number
  total_d7_active?: number
  error?: string
  dataSource?: 'appsflyer' | 'mock'
  hasData?: boolean
  baselineDays?: number
}

export interface UpdateAllBaselinesResult {
  success: boolean
  updated_count: number
  failed_count: number
  total_count: number
  results: Array<{
    product: string
    country: string
    platform?: string
    channel?: string
    baseline_roas7?: number
    baseline_ret7?: number
    status: string
    error?: string
  }>
  error?: string
}

export interface CalculateBaselineParams {
  productName: string
  countryCode: string
  platform?: string
  channel?: string
  currentDate?: Date
}

export interface CalculateBaselineFromAFParams {
  appId: string
  geo: string
  mediaSource: string
}

/**
 * Get or create baseline settings for a specific app/geo/mediaSource
 * Returns existing settings or creates with defaults if not found
 */
export async function getOrCreateBaselineSettings(params: {
  appId: string
  geo: string
  mediaSource: string
}): Promise<BaselineSettings> {
  const existing = await getBaselineSettings(params)
  if (existing) return existing

  // Create with defaults
  const defaults: NewBaselineSettings = {
    appId: params.appId,
    geo: params.geo,
    mediaSource: params.mediaSource,
    baselineDays: 180, // Default per PRD
    minSampleSize: 30,
  }

  const created = await upsertBaselineSettings(defaults)
  return created!
}

/**
 * Calculate safety baseline using real AppsFlyer cohort data
 *
 * Uses PRD 6.2.5 baseline_metrics with four-level fallback (app+geo+media_source → app+geo → app+media_source → app).
 * Window: [today - (baselineDays + 30), today - baselineDays]
 * Method: cost-weighted ROAS + install-weighted retention (no P50).
 *
 * @param params - Calculation parameters (appId, geo, mediaSource)
 * @returns Baseline calculation result with ROAS7 and RET7 values
 *
 * @example
 * ```typescript
 * const baseline = await calculateBaselineFromAF({
 *   appId: "id123456789",
 *   geo: "US",
 *   mediaSource: "googleadwords_int"
 * });
 * console.log(baseline.baseline_roas7); // e.g., 0.45
 * console.log(baseline.baseline_ret7);  // e.g., 0.38
 * ```
 */
export async function calculateBaselineFromAF(
  params: CalculateBaselineFromAFParams
): Promise<BaselineResult> {
  try {
    // Get configurable baseline window
    const settings = await getOrCreateBaselineSettings(params)

    const baseline = await getBaselineMetrics({
      appId: params.appId,
      geo: params.geo,
      mediaSource: params.mediaSource,
      daysSinceInstall: 7,
      baselineDays: settings.baselineDays,
    })

    const hasData = baseline !== null && (baseline.baselineRoas !== null || baseline.baselineRetention !== null)
    const now = new Date()
    const endDate = now.toISOString().split('T')[0]
    const startDate = new Date(now)
    startDate.setDate(startDate.getDate() - (settings.baselineDays + 30))
    const referencePeriod = `${startDate.toISOString().split('T')[0]} to ${endDate}`

    return {
      baseline_roas7: baseline?.baselineRoas ?? null,
      baseline_ret7: baseline?.baselineRetention ?? null,
      reference_period: referencePeriod,
      dataSource: 'appsflyer',
      hasData,
      baselineDays: settings.baselineDays,
    }
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : 'Unknown error'
    return {
      baseline_roas7: null,
      baseline_ret7: null,
      reference_period: '',
      error: `Failed to calculate baseline from AppsFlyer: ${errorMessage}`,
      dataSource: 'appsflyer',
      hasData: false,
    }
  }
}

/**
 * Batch update all safety baselines using AppsFlyer data
 *
 * Queries all unique app/geo/mediaSource combinations from AppsFlyer data
 * and updates their baselines.
 *
 * @returns Update results with success/failure counts
 */
export async function updateAllBaselinesFromAF(): Promise<UpdateAllBaselinesResult> {
  // Import here to avoid circular dependency
  const { getUniqueAppGeoMediaCombinations } = await import('../../db/queries-appsflyer')

  try {
    const combinations = await getUniqueAppGeoMediaCombinations()

    const results: UpdateAllBaselinesResult['results'] = []
    let updatedCount = 0
    let failedCount = 0

    for (const combo of combinations) {
      try {
        const baseline = await calculateBaselineFromAF({
          appId: combo.appId,
          geo: combo.geo,
          mediaSource: combo.mediaSource,
        })

        if (baseline.hasData) {
          results.push({
            product: combo.appId,
            country: combo.geo,
            channel: combo.mediaSource,
            baseline_roas7: baseline.baseline_roas7 ?? undefined,
            baseline_ret7: baseline.baseline_ret7 ?? undefined,
            status: 'success',
          })
          updatedCount++
        } else {
          results.push({
            product: combo.appId,
            country: combo.geo,
            channel: combo.mediaSource,
            status: 'no_data',
            error: 'Insufficient data for baseline calculation',
          })
          failedCount++
        }
      } catch (error) {
        const errorMessage = error instanceof Error ? error.message : 'Unknown error'
        results.push({
          product: combo.appId,
          country: combo.geo,
          channel: combo.mediaSource,
          status: 'error',
          error: errorMessage,
        })
        failedCount++
      }
    }

    return {
      success: failedCount === 0,
      updated_count: updatedCount,
      failed_count: failedCount,
      total_count: combinations.length,
      results,
    }
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : 'Unknown error'
    return {
      success: false,
      updated_count: 0,
      failed_count: 0,
      total_count: 0,
      results: [],
      error: `Failed to update baselines: ${errorMessage}`,
    }
  }
}

// ============================================
// LEGACY PYTHON-BASED FUNCTIONS (DEPRECATED)
// ============================================

/**
 * Run Python script and return parsed JSON output
 *
 * @deprecated Use calculateBaselineFromAF instead
 * @param scriptName - Python script filename
 * @param input - Input data to pass to Python script via stdin
 * @returns Parsed output from Python script
 */
async function runPythonScript<T>(scriptName: string, input: Record<string, unknown>): Promise<T> {
  console.warn(
    '[DEPRECATED] Using Python-based baseline calculation. Switch to calculateBaselineFromAF for real AppsFlyer data.'
  )

  return new Promise((resolve, reject) => {
    const scriptPath = path.join(process.cwd(), 'server', 'evaluation', 'python', scriptName)

    const pythonProcess = spawn('python3', [scriptPath])

    let stdoutData = ''
    let stderrData = ''

    // Collect stdout
    pythonProcess.stdout.on('data', (data) => {
      stdoutData += data.toString()
    })

    // Collect stderr
    pythonProcess.stderr.on('data', (data) => {
      stderrData += data.toString()
    })

    // Handle process exit
    pythonProcess.on('close', (code) => {
      if (code !== 0) {
        reject(new Error(`Python script exited with code ${code}\nStderr: ${stderrData}`))
        return
      }

      try {
        // Parse JSON output from stdout
        const result = JSON.parse(stdoutData)
        resolve(result)
      } catch (error) {
        reject(
          new Error(
            `Failed to parse Python output: ${error}\nOutput: ${stdoutData}\nStderr: ${stderrData}`
          )
        )
      }
    })

    // Handle process errors
    pythonProcess.on('error', (error) => {
      reject(new Error(`Failed to start Python process: ${error.message}`))
    })

    // Send input data to Python script via stdin
    pythonProcess.stdin.write(JSON.stringify(input))
    pythonProcess.stdin.end()
  })
}

/**
 * Calculate safety baseline for a specific product/country/platform/channel
 *
 * @deprecated Use calculateBaselineFromAF instead. This function uses mock data via Python.
 *
 * This calculates ROAS7 and RET7 baselines based on data from 180 days ago.
 *
 * @param params - Calculation parameters
 * @returns Baseline calculation result
 */
export async function calculateBaseline(params: CalculateBaselineParams): Promise<BaselineResult> {
  console.warn('[DEPRECATED] calculateBaseline uses mock data. Use calculateBaselineFromAF instead.')

  const input = {
    action: 'calculate',
    productName: params.productName,
    countryCode: params.countryCode,
    platform: params.platform || 'Android',
    channel: params.channel || 'Google',
    currentDate: params.currentDate?.toISOString(),
  }

  const result = await runPythonScript<BaselineResult>('baseline_calculator.py', input)
  return { ...result, dataSource: 'mock' }
}

/**
 * Batch update all safety baselines
 *
 * @deprecated Use updateAllBaselinesFromAF instead. This function uses mock data via Python.
 *
 * This should be run on the 1st of each month to update all product/country/platform/channel baselines.
 *
 * @param currentDate - Current date (optional, defaults to now)
 * @returns Update results with success/failure counts
 */
export async function updateAllBaselines(currentDate?: Date): Promise<UpdateAllBaselinesResult> {
  console.warn(
    '[DEPRECATED] updateAllBaselines uses mock data. Use updateAllBaselinesFromAF instead.'
  )

  const input = {
    action: 'update_all',
    currentDate: currentDate?.toISOString(),
  }

  return runPythonScript<UpdateAllBaselinesResult>('baseline_calculator.py', input)
}

/**
 * Re-export database functions for direct access
 */
export { getSafetyBaseline, upsertSafetyBaseline }

File: server/evaluation/python/baseline_calculator.py
#!/usr/bin/env python3
"""
A2: Safety Baseline Calculator

Calculates ROAS7 and RET7 baselines based on data from 180 days ago (6 months prior).
This baseline represents the "passing grade" for campaign performance.
"""

import sys
from datetime import datetime, timedelta
from typing import Dict, Any, List, Optional
from db_utils import get_db, format_output, read_input


class BaselineCalculator:
    """Calculate safety baselines for campaign evaluation"""

    def __init__(self):
        self.db = get_db()

    def calculate_baseline(
        self,
        product_name: str,
        country_code: str,
        platform: str = "Android",
        channel: str = "Google",
        current_date: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Calculate baseline ROAS7 and RET7 for a specific product/country/platform/channel

        Args:
            product_name: Product name (e.g., "Solitaire")
            country_code: Country code (e.g., "US")
            platform: Platform (default: "Android")
            channel: Channel (default: "Google")
            current_date: Current date in ISO format (default: today)

        Returns:
            Dictionary containing:
            - baseline_roas7: float
            - baseline_ret7: float
            - reference_period: str (e.g., "2024-06")
            - total_spend: float
            - total_revenue: float
            - total_installs: int
            - total_d7_active: int
        """
        try:
            # Parse current date
            if current_date:
                current = datetime.fromisoformat(current_date.replace('Z', '+00:00'))
            else:
                current = datetime.now()

            # Calculate reference period (6 months ago)
            reference_date = current - timedelta(days=180)
            reference_period = reference_date.strftime("%Y-%m")

            # Calculate month boundaries
            year = reference_date.year
            month = reference_date.month
            month_start = f"{year}-{month:02d}-01"

            # Calculate last day of the month
            if month == 12:
                next_month = f"{year + 1}-01-01"
            else:
                next_month = f"{year}-{month + 1:02d}-01"

            with self.db:
                # Query campaign data for the reference period
                # This is Mock Data query - will use mock_campaign_performance table
                campaign_query = """
                    SELECT
                        COALESCE(SUM(total_spend), 0) as total_spend,
                        COALESCE(SUM(total_revenue), 0) as total_revenue,
                        COALESCE(SUM(total_installs), 0) as total_installs,
                        COALESCE(SUM(d7_active_users), 0) as total_d7_active
                    FROM mock_campaign_performance
                    WHERE product_name = %s
                      AND country_code = %s
                      AND platform = %s
                      AND channel = %s
                      AND date >= %s
                      AND date < %s
                """

                results = self.db.execute_query(
                    campaign_query,
                    (product_name, country_code, platform, channel, month_start, next_month)
                )

                if not results or len(results) == 0:
                    return {
                        "error": f"No data found for {product_name}/{country_code}/{platform}/{channel} in {reference_period}",
                        "baseline_roas7": None,
                        "baseline_ret7": None,
                        "reference_period": reference_period
                    }

                data = results[0]
                total_spend = float(data['total_spend'])
                total_revenue = float(data['total_revenue'])
                total_installs = int(data['total_installs'])
                total_d7_active = int(data['total_d7_active'])

                # Calculate baselines
                baseline_roas7 = (total_revenue / total_spend) if total_spend > 0 else 0.0
                baseline_ret7 = (total_d7_active / total_installs) if total_installs > 0 else 0.0

                return {
                    "baseline_roas7": round(baseline_roas7, 4),
                    "baseline_ret7": round(baseline_ret7, 4),
                    "reference_period": reference_period,
                    "total_spend": total_spend,
                    "total_revenue": total_revenue,
                    "total_installs": total_installs,
                    "total_d7_active": total_d7_active
                }

        except Exception as e:
            return {
                "error": str(e),
                "baseline_roas7": None,
                "baseline_ret7": None
            }

    def upsert_baseline(
        self,
        product_name: str,
        country_code: str,
        platform: str,
        channel: str,
        baseline_roas7: float,
        baseline_ret7: float,
        reference_period: str
    ) -> bool:
        """
        Insert or update baseline in safety_baseline table

        Args:
            product_name: Product name
            country_code: Country code
            platform: Platform
            channel: Channel
            baseline_roas7: Calculated ROAS7 baseline
            baseline_ret7: Calculated RET7 baseline
            reference_period: Reference period (e.g., "2024-06")

        Returns:
            True if successful, False otherwise
        """
        try:
            with self.db:
                query = """
                    INSERT INTO safety_baseline (
                        product_name, country_code, platform, channel,
                        baseline_roas7, baseline_ret7, reference_period, last_updated
                    )
                    VALUES (%s, %s, %s, %s, %s, %s, %s, NOW())
                    ON CONFLICT (product_name, country_code, platform, channel)
                    DO UPDATE SET
                        baseline_roas7 = EXCLUDED.baseline_roas7,
                        baseline_ret7 = EXCLUDED.baseline_ret7,
                        reference_period = EXCLUDED.reference_period,
                        last_updated = NOW()
                """

                return self.db.execute_update(
                    query,
                    (product_name, country_code, platform, channel,
                     baseline_roas7, baseline_ret7, reference_period)
                )

        except Exception as e:
            print(f"Upsert baseline error: {e}", file=sys.stderr, flush=True)
            return False

    def update_all_baselines(self, current_date: Optional[str] = None) -> Dict[str, Any]:
        """
        Batch update baselines for all product/country/platform/channel combinations

        This should be run on the 1st of each month to update all baselines.

        Args:
            current_date: Current date in ISO format (default: today)

        Returns:
            Dictionary containing:
            - success: bool
            - updated_count: int
            - failed_count: int
            - results: List of update results
        """
        try:
            with self.db:
                # Query all unique combinations from mock data
                combo_query = """
                    SELECT DISTINCT
                        product_name,
                        country_code,
                        platform,
                        channel
                    FROM mock_campaign_performance
                    ORDER BY product_name, country_code, platform, channel
                """

                combinations = self.db.execute_query(combo_query)

                updated_count = 0
                failed_count = 0
                results = []

                for combo in combinations:
                    # Calculate baseline for this combination
                    baseline = self.calculate_baseline(
                        product_name=combo['product_name'],
                        country_code=combo['country_code'],
                        platform=combo['platform'],
                        channel=combo['channel'],
                        current_date=current_date
                    )

                    if 'error' not in baseline and baseline['baseline_roas7'] is not None:
                        # Upsert to database
                        success = self.upsert_baseline(
                            product_name=combo['product_name'],
                            country_code=combo['country_code'],
                            platform=combo['platform'],
                            channel=combo['channel'],
                            baseline_roas7=baseline['baseline_roas7'],
                            baseline_ret7=baseline['baseline_ret7'],
                            reference_period=baseline['reference_period']
                        )

                        if success:
                            updated_count += 1
                            results.append({
                                "product": combo['product_name'],
                                "country": combo['country_code'],
                                "platform": combo['platform'],
                                "channel": combo['channel'],
                                "baseline_roas7": baseline['baseline_roas7'],
                                "baseline_ret7": baseline['baseline_ret7'],
                                "status": "updated"
                            })
                        else:
                            failed_count += 1
                            results.append({
                                "product": combo['product_name'],
                                "country": combo['country_code'],
                                "status": "failed"
                            })
                    else:
                        failed_count += 1
                        results.append({
                            "product": combo['product_name'],
                            "country": combo['country_code'],
                            "status": "no_data",
                            "error": baseline.get('error', 'Unknown error')
                        })

                return {
                    "success": True,
                    "updated_count": updated_count,
                    "failed_count": failed_count,
                    "total_count": len(combinations),
                    "results": results
                }

        except Exception as e:
            return {
                "success": False,
                "error": str(e),
                "updated_count": 0,
                "failed_count": 0
            }


def main():
    """Main entry point for CLI usage"""
    # Read input from stdin (sent by TypeScript wrapper)
    input_data = read_input()

    calculator = BaselineCalculator()

    action = input_data.get('action', 'calculate')

    if action == 'calculate':
        # Calculate baseline for specific combination
        result = calculator.calculate_baseline(
            product_name=input_data.get('productName', 'Solitaire'),
            country_code=input_data.get('countryCode', 'US'),
            platform=input_data.get('platform', 'Android'),
            channel=input_data.get('channel', 'Google'),
            current_date=input_data.get('currentDate')
        )
    elif action == 'update_all':
        # Batch update all baselines
        result = calculator.update_all_baselines(
            current_date=input_data.get('currentDate')
        )
    else:
        result = {"error": f"Unknown action: {action}"}

    # Output result as JSON to stdout
    format_output(result)


if __name__ == "__main__":
    main()
